sphinx_needs/utils.py

```diff
@@ -542,110 +541,85 @@ def match_string_link(
-            check_definition = variant_pattern.search(variant_definition)
-            if check_definition:
-                variants_in_option = True
-                # Separate variant definition from value to use for the option
-                filter_string, output, _ = re.split(
-                    r"(:[\w':.\-\" ]+)$", variant_definition
-                )
-                filter_string = re.sub(r"^\[|[:\]]$", "", filter_string)
-                filter_string = needs_variants.get(filter_string, filter_string)
-                try:
-                    # https://docs.python.org/3/library/functions.html?highlight=compile#compile
-                    filter_compiled = compile(filter_string, "<string>", "eval")
-                    # Set filter_context as globals and not only locals in eval()!
-                    # Otherwise, the vars not be accessed in list comprehensions.
-                    if filter_compiled:
-                        eval_result = bool(eval(filter_compiled, filter_context))
-                    else:
-                        eval_result = bool(eval(filter_string, filter_context))
-                    # First matching variant definition defines the output
-                    if eval_result:
-                        no_variants_in_option = False
-                        return output.lstrip(":")
-                except Exception as e:
-                    logger.warning(
-                        f'There was an error in the filter statement: "{filter_string}". '
-                        f"Error Msg: {e} [needs]",
-                        type="needs",
-                    )
-            else:
-                no_variants_in_option = True
-
-        if no_variants_in_option and not variants_in_option:
-            return None
-
-        # If no variant-rule is True, set to last, variant-free option. If this does not exist, set to None.
-        defaults_to = variant_definitions[-1]
-        if variants_in_option and variant_pattern.search(defaults_to):
-            return None
-        return re.sub(r"[;,] ", "", defaults_to)
-
-    split_pattern = r"([\[\]]{1}[\w=:'. \-\"]+[\[\(\{]{1}[\w=,.': \-\"]*[\]\)\}]{1}[\[\]]{1}:[\w.\- ]+)|([\[\]]{1}[\w=:.'\-\[\] \"]+[\[\]]{1}:[\w.\- ]+)|([\w.: ]+[,;]{1})"
-    variant_rule_pattern = r"^[\w'=,:.\-\"\[\] ]+:[\w'=:.\-\"\[\] ]+$"
-    variant_splitting = re.compile(split_pattern)
-    variant_rule_matching = re.compile(variant_rule_pattern)
-
-    # Handling multiple variant definitions
-    if isinstance(option_value, str):
-        multiple_variants: list[str] = variant_splitting.split(rf"""{option_value}""")
-        multiple_variants = [
+    options_list: list[str]
+    if isinstance(options, str):
+        options_list = re.split(
+            r"([\[\]]{1}[\w=:'. \-\"]+[\[\(\{]{1}[\w=,.': \-\"]*[\]\)\}]{1}[\[\]]{1}:[\w.\- ]+)|([\[\]]{1}[\w=:.'\-\[\] \"]+[\[\]]{1}:[\w.\- ]+)|([\w.: ]+[,;]{1})",
+            rf"""{options}""",
```

Comment:
Should this regex be precompiled?

Otherwise, I guess it gets recompiled with every need-option?!

Reply:
Nope, Python caches every compiled regexes: https://github.com/python/cpython/blob/8e6321efd72d12263398994e59c5216edcada3c0/Lib/re/__init__.py#L330

so in fact it was inefficient to be re-compiling on every call to this function, and this should improve perfomance!